internal/pkg/instance/logger.go

```diff
@@ -114,39 +111,68 @@ func (l *Logger) scanOutput(data []byte, atEOF bool) (advance int, token []byte,
 // NewWriter create a new pipe pair for corresponding stream.
 func (l *Logger) NewWriter(stream string, dropCRNL bool) *io.PipeWriter {
 	reader, writer := io.Pipe()
-	go l.scan(stream, reader, writer, dropCRNL)
+	closer := l.scan(stream, reader, writer, dropCRNL)
+	l.Lock()
+	l.closers = append(l.closers, closer)
+	l.Unlock()
 	return writer
 }
 
-func (l *Logger) scan(stream string, pr *io.PipeReader, pw *io.PipeWriter, dropCRNL bool) {
+func (l *Logger) scan(stream string, pr *io.PipeReader, pw *io.PipeWriter, dropCRNL bool) closer {
 	r := strings.NewReplacer("\r", "\\r", "\n", "\\n")
 	scanner := bufio.NewScanner(pr)
 	if !dropCRNL {
 		scanner.Split(l.scanOutput)
 	}
 
-	for scanner.Scan() {
-		l.fileMutex.Lock()
-		if !dropCRNL {
-			fmt.Fprint(l.file, l.formatter(stream, r.Replace(scanner.Text())))
-		} else {
-			fmt.Fprint(l.file, l.formatter(stream, scanner.Text()))
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	go func() {
+		for scanner.Scan() {
+			l.Lock()
```

Comment:
this is the bit where I'm a bit confused. What's this lock protecting? The client calling ReOpenFile while this is running? A comment to that effect would be needed here. The other thing this could be protecting is having multiple writers (with NewWriter) trying to write at the same time. Same thing: this needs a comment.

Also, if that's what's going on here, note that calling ReOpenFile might end up in a situation where the file handle is no longer valid (l.file was closed, a new file handle could not be obtained). After `l.Lock()` there should be a check like `if l.file == nil { break }`.

Reply:
Will add